### PR TITLE
Feature: Add dataprocessor.

### PR DESCRIPTION
**Pull request by joernweissenborn**
Adds a (chainable) PreProcessingPipeline with (initial) support for CorrectBaselineAverage and CorrectBaselineValue.

### CODE REVIEW

File: glotaran/io/preprocessor/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        Returns
+        -------
+        PreProcessingPipeline
+        """
+        self._push_action(CorrectBaselineValue(value=value))
+        return self
+
+    def correct_baseline_average(self, interval: dict[str, slice]) -> PreProcessingPipeline:
+        """Correct a dataset by subtracting an average of the data.
```
Comment (jsnel) [inline glotaran/io/preprocessor/pipeline.py:68]:
```suggestion
        """Correct a dataset by subtracting the average over a part of the data.
```

File: glotaran/io/preprocessor/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+        Parameters
+        ----------
+        interval: dict[str, slice]
+            The intervals to average.
```
Comment (jsnel) [inline glotaran/io/preprocessor/pipeline.py:73]:
```suggestion
            The intervals to calculate the average over to subtract from the data.
```

File: glotaran/io/preprocessor/preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            The data to process.
+
+        Returns
+        -------
+        xr.DataArray
+        """
+        selection = data
+        for axis, interval in self.interval.items():
+            selection = selection.sel({axis: interval})
+        return data - (selection.sum() / selection.size)
```
Comment (jsnel) [inline glotaran/io/preprocessor/preprocessor.py:78]: This code needs some revision, it suggests that it works with more than one intervals, but it doesn't.

Suppose an time axis [0,100], are you intending it to work, [0,10] and [90,100] in one operations, or only [0,10] and then [90,100] in two operations. It's not clear, and there's no tests for this either.

**Review by jsnel — CHANGES_REQUESTED** [review]
Nice start, but I'd like to see a bit more meaning full tests. 
The CorrectBaselineValue method is ok, just needs more tests.
The CorrectBaseLineAverage case needs revision, and more extensive tests.

File: glotaran/io/preprocessor/preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        ----------
+        data: xr.DataArray
+            The data to process.
+
+        Returns
+        -------
+        xr.DataArray
+        """
+        selection = data.sel(self.selection or {}).drop_sel(self.exclude or {})
+        return data - (selection.sum() / selection.size)
```
Comment (s-weigand) [inline glotaran/io/preprocessor/preprocessor.py:77]:
```suggestion
        return data - selection.mean()
```

File: glotaran/io/preprocessor/preprocessor.py (hunk 3 above)
Comment (joernweissenborn) [reply glotaran/io/preprocessor/preprocessor.py]: I also added `exclude`, if you eg want to average only the edges.

File: glotaran/io/preprocessor/test/test_preprocessor.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+def test_correct_baseline_value():
+    pl = PreProcessingPipeline()
+    pl.correct_baseline_value(1)
+    data = xr.DataArray([[1]])
+    result = pl.apply(data)
+    assert result == data - 1
+
+
+@pytest.mark.parametrize("indexer", (slice(0, 2), [0, 1]))
+def test_correct_baseline_average(indexer: slice | list[int]):
```
Comment (s-weigand) [inline glotaran/io/preprocessor/test/test_preprocessor.py:16]: Please also add a test for exclude and one that shows a combination of selection and exclude.

File: glotaran/io/preprocessor/preprocessor.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        """
+        return data - self.value
+
+
+class CorrectBaselineAverage(PreProcessor):
+    """Corrects a dataset by subtracting the average over a part of the data."""
+
+    action: Literal["baseline-average"] = "baseline-average"
+    selection: dict[str, slice | list[int] | int] | None = None
+    exclude: dict[str, slice | list[int] | int] | None = None
```
Comment (s-weigand) [inline glotaran/io/preprocessor/preprocessor.py:62]: For consistency, we should name the arguments
- select
- exclude
Or
- selection
- exclusion

**Review by s-weigand — CHANGES_REQUESTED** [review]
Looks like a good start, just got my usual nitpicks.

I guess for now there be no other way to compose pipelines than copypasting a file together or combining the action attributes.

```py
compose_pipeline = PreProcessingPipeline(pl1.actions[:-1] + pl2.actions[-2:])
```

File: glotaran/io/preprocessor/__init__.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+"""Tools for data pre-processing."""
+from glotaran.io.preprocessor.pipeline import PreProcessingPipeline
```
Comment (s-weigand) [inline glotaran/io/preprocessor/__init__.py:2]: `CorrectBaselineValue` and `CorrectBaselineAverage` also be reexported from here for ease of use.

**Review by s-weigand — CHANGES_REQUESTED** [review]
Sound something else (sorry for not catching it before 😢)

File: glotaran/io/preprocessor/__init__.py (hunk 7 above)
Comment (joernweissenborn) [reply glotaran/io/preprocessor/__init__.py:2]: Nobody should use them directly, it makes no sense. They are considered internal.

Comment (s-weigand) [reply glotaran/io/preprocessor/__init__.py:2]: Yeah nvm I'm just tired

**Review by s-weigand — APPROVED** [review]
Looks good to me now, thanks for adding the preprocessing. 🎉

**Review by jsnel — DISMISSED** [review]
It may not be the most intuitive way of doing preprocessing, but it works (for the limited cases currently defined), and it doesn't conflict with anything, so I'm ok with merging it in. 

I do think improvements are possible, but that can be tackled in a future PR.

File: glotaran/io/preprocessor/preprocessor.py (hunk 3 above)
Comment (jsnel) [reply glotaran/io/preprocessor/preprocessor.py]: The code is more correct and clear now than before.
```python
return data - data.sel(self.select or {}).drop_sel(self.exclude or {}).mean()
```

File: glotaran/io/preprocessor/pipeline.py (hunk 1 above)
Comment (jsnel) [reply glotaran/io/preprocessor/pipeline.py]: The suggestions were incorportated.

File: glotaran/io/preprocessor/pipeline.py (hunk 2 above)
Comment (jsnel) [reply glotaran/io/preprocessor/pipeline.py]: The suggestions were incorporated.

File: glotaran/io/preprocessor/test/test_preprocessor.py (hunk 5 above)
Comment (jsnel) [reply glotaran/io/preprocessor/test/test_preprocessor.py:16]: We'll get back to it in a future PR ;)

**Review by jsnel — APPROVED** [review]
Re-approve changes following dropped commit.

**Pull request merged**